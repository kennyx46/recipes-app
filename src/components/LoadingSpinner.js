import React from 'react';
import Container from 'react-bootstrap/Container';
import Spinner from 'react-bootstrap/Spinner';

const LoadingSpinner = () => {
	return (
		<Container className="text-center mt-4">
		    <Spinner size="lg" animation="border"/>
		</Container>
	);
};

export default LoadingSpinner;
