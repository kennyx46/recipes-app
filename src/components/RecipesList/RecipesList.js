import React, { Component } from 'react';
import Card from 'react-bootstrap/Card';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Alert from 'react-bootstrap/Alert';
import LoadingSpinner from '../LoadingSpinner';

import './RecipesList.css';

export default class RecipesList extends Component {

    componentDidMount() {
        this.props.getRecipes();
    }

    renderRecipe = (recipe) => {
        return (
            <Col xs={12} md={6} lg={4} key={recipe.id}>
                <Card className="mb-4" onClick={() => this.props.push(`/recipes/${recipe.id}`)}>
                    <Card.Img variant="top" src={recipe.image} />
                    <Card.Header>{recipe.title}</Card.Header>
                </Card>
            </Col>
        );
    }

    render() {
        const { isLoading, recipes, error } = this.props;

        if (isLoading) {
            return (
                <LoadingSpinner/>
            );
        }

        return (
            <Container>
                <h1>Recipes</h1>
                { error && <Alert variant='danger'>Someting went wrong. Please try again later</Alert> }
                <Row>
                    {recipes.map((this.renderRecipe))}
                </Row>
            </Container>
        );
    }
}
