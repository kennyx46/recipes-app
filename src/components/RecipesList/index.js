import { connect } from 'react-redux';
import { getRecipes } from '../../data/recipes/actions';
import { getRecipesItems, getRecipesError, getRecipesIsLoading } from '../../data/recipes/selectors';
import RecipesList from './RecipesList';
import { push } from 'connected-react-router';

const mapStateToProps = (state) => ({
    recipes: getRecipesItems(state),
    isLoading: getRecipesIsLoading(state),
    error: getRecipesError(state),
});

const mapDispatchToProps = {
    getRecipes,
    push
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(RecipesList);
