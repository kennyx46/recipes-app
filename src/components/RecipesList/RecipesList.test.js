import React from 'react';
import { shallow } from 'enzyme';
import Card from 'react-bootstrap/Card';
import LoadingSpinner from '../LoadingSpinner';

import RecipesList from './RecipesList';

describe('<RecipesList>', () => {
    const defaultProps = Object.freeze({
        push: jest.fn(),
        recipes: [],
        isLoading: false,
        getRecipes: jest.fn(),
    });

    it('renders component', () => {
        const wrapper = shallow(
            <RecipesList {...defaultProps }/>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('shows loading if loading in progress', () => {
        const props = { ...defaultProps, isLoading: true };
        const recipesList = shallow(<RecipesList {...props} />);
        expect(recipesList.exists(LoadingSpinner)).toBeTruthy();
    });

    it('renders <Card> for every recipe', () => {
        const sampleRecipe = { id: '1', title: 'recipe One' };
        const props = { ...defaultProps, recipes: [ sampleRecipe, { ...sampleRecipe, id: '2' }, { ...sampleRecipe, id: '3' } ] };
        const recipesList = shallow(<RecipesList {...props} />);
        expect(recipesList.find(Card)).toHaveLength(3);
    });


    it('navigate to recipe details on click', () => {
        const sampleRecipe = { id: '1', title: 'recipe One' };
        const props = { ...defaultProps, recipes: [ sampleRecipe ] };
        const recipesList = shallow(<RecipesList {...props} />);

        recipesList.find(Card).simulate('click');

        expect(defaultProps.push).toHaveBeenCalledWith(`/recipes/${sampleRecipe.id}`);
    });

});
