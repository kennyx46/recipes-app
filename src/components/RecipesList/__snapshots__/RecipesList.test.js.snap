// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecipesList> renders component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RecipesList
    getRecipes={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    isLoading={false}
    push={[MockFunction]}
    recipes={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <h1>
          Recipes
        </h1>,
        undefined,
        <ForwardRef
          noGutters={false}
        />,
      ],
      "fluid": false,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Recipes",
        },
        "ref": null,
        "rendered": "Recipes",
        "type": "h1",
      },
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [],
          "noGutters": false,
        },
        "ref": null,
        "rendered": Array [],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "noGutters": false,
          },
          "render": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "fluid": false,
      },
      "displayName": "Container",
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <h1>
            Recipes
          </h1>,
          undefined,
          <ForwardRef
            noGutters={false}
          />,
        ],
        "fluid": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Recipes",
          },
          "ref": null,
          "rendered": "Recipes",
          "type": "h1",
        },
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [],
            "noGutters": false,
          },
          "ref": null,
          "rendered": Array [],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "noGutters": false,
            },
            "render": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "fluid": false,
        },
        "displayName": "Container",
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
